### DISCUSSION TITLE
Dynamic submenus possible?

### DISCUSSION THREAD

**Discussion by dremin** [twilio-labs/paste#3286] [Q&A] [answered]
Hi! I have some data that I am attempting to dynamically generate a menu and submenus for (the data is divided into categories, and I'd like a submenu per category). Unfortunately, I keep running into issues when trying to open any of the said submenus when doing this--sometimes freeze, sometimes 'maximum update depth' errors, sometimes nothing happens. Is what I am attempting possible? [Here is my non-working code](https://codesandbox.io/s/bold-wildflower-rycyck?file=/src/index.tsx).

**Comment by nkrantz** [comment, marked as answer]
Hi @dremin - should be possible! [We forked your Codesandbox to demonstrate the behavior](https://codesandbox.io/s/jolly-architecture-8ksmgn?file=/src/index.tsx). Let us know if you have any questions. Thanks!

**Reply by dremin** [reply]
Thanks @nkrantz! That works in the sandbox, but unfortunately the additional prop on `MenuItem` upsets TypeScript, which makes that not a usable solution for us in its current state. To work around this I created a new component that wraps MenuItem, but now I am back to the previous broken behavior. I feel like I am missing something subtle about why the version you provided works 😄 [Here's an updated sandbox with the wrapped MenuItem](https://codesandbox.io/s/affectionate-burnell-csh56j?file=/src/CategorySubMenuItem.tsx), would you be able to help explain what is causing this to not work? Thanks!

**Reply by TheSisb** [reply]
The version provided by Nora works simply for the fact that she moved the looped inline-component into a functional component. From your original version, you're calling the menu hook inside a loop, inside an inline-component: 
<img width="570" alt="image" src="https://github.com/twilio-labs/paste/assets/1592327/e6d4ee44-16f7-4fdc-a9f8-280f60a785f5">

In your new example, you're doing the same thing again:
<img width="594" alt="image" src="https://github.com/twilio-labs/paste/assets/1592327/1c5a2415-8c9b-4458-a727-4174f4c9cc16">
You can review the rules of hooks here: https://legacy.reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level


--

Finally, here's a fully cleaned up version: https://codesandbox.io/s/great-heisenberg-d2rvph?file=/src/Menu/SubMenu.tsx

**Reply by dremin** [reply]
@TheSisb Thank you for the explanation 🙏 That makes sense and I have now successfully implemented this. Thanks to both you and @nkrantz for the help, I greatly appreciate it!
